### PR TITLE
Wigglesort

### PR DESCRIPTION
**Pull request by BranAndSceolan**
### **Describe your change:**
Fixes:   #3023

- [x] Add an algorithm?
- [ ] Fix a bug or typo in an existing algorithm?
- [ ] Documentation change?

### **Checklist:**

- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Java/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized.
- [x] I know that pull requests will not be merged if they fail the automated tests.
- [x] This PR only changes one algorithm file. To ease review, please open separate PRs for separate algorithms.
- [x] All new Java files are placed inside an existing directory.
- [x] All filenames are in all uppercase characters with no spaces or dashes.
- [x] All functions and variable names follow Java naming conventions.
- [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
- [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: src/main/java/com/thealgorithms/sorts/WiggleSort.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
+package com.thealgorithms.sorts;
+
+import java.util.Arrays;
+
+import static com.thealgorithms.maths.Ceil.ceil;
+import static com.thealgorithms.maths.Floor.floor;
+import static com.thealgorithms.searches.QuickSelect.select;
+
+/**
+ * A wriggle sort implementation based on John L.s' answer in
```
Comment (siriak) [inline src/main/java/com/thealgorithms/sorts/WiggleSort.java:10]:
```suggestion
 * A wiggle sort implementation based on John L.s' answer in
```

File: src/main/java/com/thealgorithms/sorts/WiggleSort.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+    private <T extends Comparable<T>> T[] wiggleSort(T[] sortThis) {
+        // find the median using quickSelect (if the result isn't in the array, use the next greater value)
+        T median;
+
+        if(sortThis.length % 2 == 0) {
+            median = select(Arrays.<T>asList(sortThis), (int) (sortThis.length / 2.0));
+        } else {
+            median = select(Arrays.<T>asList(sortThis), (int) floor(sortThis.length / 2.0));
+        }
```
Comment (siriak) [inline src/main/java/com/thealgorithms/sorts/WiggleSort.java:56]: These are the same, you could have used floor for both branches

File: src/main/java/com/thealgorithms/sorts/WiggleSort.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+        System.out.println(median);
```
Comment (siriak) [inline src/main/java/com/thealgorithms/sorts/WiggleSort.java:58]: Please remove stdout output

File: src/main/java/com/thealgorithms/sorts/WiggleSort.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+            }
+            if (numMedians > ceil(sortThis.length / 2.0)) {
+                throw new IllegalArgumentException("No more than half the number of values may be the same.");
+            }
+        }
+
+        triColorSort(sortThis, median);
+        return sortThis;
+    }
+    public static void main(String[] args) {
```
Comment (siriak) [inline src/main/java/com/thealgorithms/sorts/WiggleSort.java:73]: Please add automated tests with @Test instead of main()

**Review by siriak — CHANGES_REQUESTED** [review]

**Comment by BranAndSceolan** [comment]
Thank you for reviewing, I have made the changes you requested.

**Review by siriak — APPROVED** [review]
LGTM, thanks!

**Pull request merged**